third_party/pcsc-lite/naclport/server_clients_management/src/client_request_processor.cc

```diff
@@ -223,17 +213,63 @@ void PcscLiteClientRequestProcessor::AsyncProcessRequest(
     std::shared_ptr<PcscLiteClientRequestProcessor> request_processor,
     RemoteCallRequestPayload request,
     RequestReceiver::ResultCallback result_callback) {
-  // This is paired with the call to `LeaveConcurrencyCheckScope()` in
-  // `ProcessRequest()`. We enter the scope here in order to be able to catch
-  // concurrency violations even when they occur so fast that the thread created
-  // below didn't have chance to start and do any work.
-  request_processor->EnterConcurrencyCheckScope(request.function_name);
-
   std::thread(&PcscLiteClientRequestProcessor::ProcessRequest,
               request_processor, std::move(request), result_callback)
       .detach();
 }
 
+PcscLiteClientRequestProcessor::ScopedConcurrencyGuard::ScopedConcurrencyGuard(
+    const std::string& function_name,
+    SCARDCONTEXT s_card_context,
+    SCARDHANDLE s_card_handle,
+    PcscLiteClientRequestProcessor& owner)
+    : owner_(owner),
+      function_name_(function_name),
+      s_card_context_(s_card_context) {
+  s_card_context_ =
+      s_card_handle
+          ? owner_.s_card_handles_registry_.FindContextByHandle(s_card_handle)
+          : s_card_context;
+  if (!s_card_context_) {
+    return;
+  }
+
+  std::string concurrent_functions_dump;
+  {
+    const std::unique_lock<std::mutex> lock(owner_.mutex_);
+    std::multiset<std::string>& running_functions =
+        owner_.context_to_running_functions_[s_card_context_];
+    concurrent_functions_dump =
+        JoinStrings(running_functions, /*separator=*/", ");
+    running_functions.insert(function_name_);
+  }
+
+  if (!concurrent_functions_dump.empty()) {
+    GOOGLE_SMART_CARD_LOG_WARNING
+        << owner_.logging_prefix_
+        << "Client violates threading: concurrent calls of " << function_name_
+        << ", " << concurrent_functions_dump
+        << " against the same SCARDCONTEXT. Future releases of Smart Card "
+           "Connector will forbid this: every call referring to some "
+           "SCARDCONTEXT must come after the previous one completed.";
+  }
+}
+
+PcscLiteClientRequestProcessor::ScopedConcurrencyGuard::
+    ~ScopedConcurrencyGuard() {
+  if (!s_card_context_) {
+    return;
+  }
+  const std::unique_lock<std::mutex> lock(owner_.mutex_);
+  auto iter = owner_.context_to_running_functions_.find(s_card_context_);
+  GOOGLE_SMART_CARD_CHECK(iter != owner_.context_to_running_functions_.end());
+  std::multiset<std::string>& running_functions = iter->second;
+  running_functions.erase(running_functions.find(function_name_));
```

Comment:
This would erase all entries with the given name, even though only one should be erased (disregarding that we are already in a non-supported state at this point)

Reply:
Hmm I think it only delete exactly one item because we pass an iterator. WDYT?

Reply:
You are right, I did not see the find() call in there